python/utils/gpu_metric_poller.py

```diff
@@ -0,0 +1,168 @@
+    """
+    def __init__(self, *args, **kwargs):
+        self.__stop = False
+        super().__init__(*args, **kwargs)
+        self.maxGpuUtil = 0
+        self.maxGpuMemUsed = 0
+
+    @staticmethod
+    def __waitForInput(fd):
+        # assume non-blocking fd
+        while True:
+            if not fd.closed:
+                line = fd.readline()
+                if line:
+                    return line
+            else:
+                break
+        return None
+
+    @staticmethod
+    def __writeToPipe(fd, strToWrite):
+        fd.write(strToWrite)
+        fd.flush()
+
+    def __runParentLoop(self, readFileNo, writeFileNo):
+        parentReadPipe = os.fdopen(readFileNo)
+        parentWritePipe = os.fdopen(writeFileNo, "w")
+
+        self.__writeToPipe(parentWritePipe, "1")
+        gpuMetricsStr = self.__waitForInput(parentReadPipe)
+        while True:
+            # FIXME: this assumes the input received is perfect!
+            (memUsed, gpuUtil) = [int(x) for x in
+                                  gpuMetricsStr.strip().split()]
+
+            if memUsed > self.maxGpuMemUsed:
+                self.maxGpuMemUsed = memUsed
+            if gpuUtil > self.maxGpuUtil:
+                self.maxGpuUtil = gpuUtil
+
+            if not self.__stop:
+                self.__writeToPipe(parentWritePipe, "1")
+            else:
+                self.__writeToPipe(parentWritePipe, "0")
+                break
+            gpuMetricsStr = self.__waitForInput(parentReadPipe)
+
+        parentReadPipe.close()
+        parentWritePipe.close()
+
+    def __runChildLoop(self, readFileNo, writeFileNo):
+        childReadPipe = os.fdopen(readFileNo)
+        childWritePipe = os.fdopen(writeFileNo, "w")
+
+        smi.nvmlInit()
+        # hack - get actual device ID somehow
+        devObj = smi.nvmlDeviceGetHandleByIndex(0)
+        memObj = smi.nvmlDeviceGetMemoryInfo(devObj)
+        utilObj = smi.nvmlDeviceGetUtilizationRates(devObj)
+        initialMemUsed = memObj.used
+        initialGpuUtil = utilObj.gpu
+
+        controlStr = self.__waitForInput(childReadPipe)
+        while True:
+            memObj = smi.nvmlDeviceGetMemoryInfo(devObj)
+            utilObj = smi.nvmlDeviceGetUtilizationRates(devObj)
+
+            memUsed = memObj.used - initialMemUsed
+            gpuUtil = utilObj.gpu - initialGpuUtil
+
+            if controlStr.strip() == "1":
+                self.__writeToPipe(childWritePipe, "%s %s\n"
+                                   % (memUsed, gpuUtil))
+            elif controlStr.strip() == "0":
+                break
+            controlStr = self.__waitForInput(childReadPipe)
+
+        smi.nvmlShutdown()
+        childReadPipe.close()
+        childWritePipe.close()
+
+    def run(self):
+        (parentReadPipeFileNo, childWritePipeFileNo) = os.pipe2(os.O_NONBLOCK)
+        (childReadPipeFileNo, parentWritePipeFileNo) = os.pipe2(os.O_NONBLOCK)
```

Comment:
parentReadPipeFileNo and childReadPipeFileNo will be the same number based on the syntax of this call.  so why not just have readPipeNo and not have the second call?

Reply:
Those calls create a pair of one-way pipes, each with "ends" that need to be used by the parent and the child. I create two of them for the two-way communication (so the parent can issue a poll and the child can send the numbers back), and each end is a unique file descriptor (I checked in the debugger):
```
(Pdb) p (parentReadPipeFileNo, childWritePipeFileNo)
(4, 5)
(Pdb) p (childReadPipeFileNo, parentWritePipeFileNo)
(6, 7)
```
unless, were you thinking the two-way communication isn't necessary? ...which in thinking about it now, I can just have the child constantly stream back numbers with only a single pipe.

Reply:
I was thinking that each call to os.pipe2 would return the same IDs.